### PR TITLE
JWS-based Notation CLI

### PR DESCRIPTION
**Pull request by shizhMSFT**
Change logs:

- Merged package `notation-go-lib/registry` to package `notation/pkg/registry`.
- Updated the JWT signature format to JWS signature format by updating the notation go library.
  - Dependency change: https://github.com/notaryproject/notation-go-lib/pull/11.
  - Timestamping is integrated.

### CODE REVIEW

File: pkg/config/map.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+// CertificateReference is a named file path.
+type CertificateReference struct {
+	Name string `json:"name"`
+	Path string `json:"path"`
+}
+
+// CertificateMap is a set of CertificateReference indexed by name.
+// The overall performance is O(n) while the order of entries is persevered.
+type CertificateMap []CertificateReference
```
Comment (priteshbandi) [inline pkg/config/map.go:61]: Why do we need both CertificateMap and KeyMap? Can we reuse KeyMap to store certificates with key as nil.

File: pkg/signature/jws.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+// NewSignerFromFiles creates a signer from key, certificate files
+func NewSignerFromFiles(keyPath, certPath string) (*jws.Signer, error) {
+	if keyPath == "" {
+		return nil, errors.New("key path not specified")
+	}
+	if certPath == "" {
+		return nil, errors.New("certificate path not specified")
+	}
```
Comment (priteshbandi) [inline pkg/signature/jws.go:20]: Should we add blank string check? strings.TrimSpace(keyPath) == "" or we are already doing that some where before this function is called?

File: pkg/signature/jws.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		bundledCerts, err := cryptoutil.ReadCertificateFile(path)
+		if err != nil {
+			return nil, err
+		}
+		for _, cert := range bundledCerts {
+			roots.AddCert(cert)
+		}
+	}
+	verifier := jws.NewVerifier()
+	verifier.VerifyOptions.Roots = roots
```
Comment (priteshbandi) [inline pkg/signature/jws.go:64]: Why do we care about root or non-root cert in verifier?

File: go.mod [hunk 4]
```diff
@@ -2,26 +2,27 @@ module github.com/notaryproject/notation
 
 go 1.17
 
+replace github.com/notaryproject/notation-go-lib => github.com/shizhMSFT/notation-go-lib v0.0.0-20210928070605-9b2fdcfba155
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply go.mod:5]: Definitely!

File: cmd/notation/common.go [hunk 5]
```diff
@@ -3,38 +3,38 @@ package main
...
 	}
-	localFlag = &cli.BoolFlag{
+	flagLocal = &cli.BoolFlag{
 		Name:    "local",
 		Aliases: []string{"l"},
 		Usage:   "reference is a local file",
 	}
-	signatureFlag = &cli.StringSliceFlag{
+	flagSignature = &cli.StringSliceFlag{
 		Name:      "signature",
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply cmd/notation/common.go:38]: That's the go language specific stuff.

```go
var (
   privateVariable  int // camelCase variables are only accessible within the package.
   ExportedVariable int // PascalCase variables are exported and can be accessed anywhere.
)
```

File: internal/version/version.go [hunk 6]
```diff
@@ -2,7 +2,7 @@ package version
 
 var (
 	// Version shows the current notation version, optionally with pre-release.
-	Version = "0.5.3"
+	Version = "0.6.1"
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply internal/version/version.go:5]: Currently, I am tracking it manually. It follows SemVer2.

File: internal/cmd/signer.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/notaryproject/notation/pkg/config"
+	"github.com/notaryproject/notation/pkg/signature"
+	"github.com/urfave/cli/v2"
+)
+
+// GetSigner returns a signer according to the CLI context.
+func GetSigner(ctx *cli.Context) (notation.Signer, error) {
+	// read paths of the signing key and its corresponding cert.
+	var keyPath, certPath string
+	if path := ctx.String(FlagKeyFile.Name); path != "" {
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply internal/cmd/signer.go:17]: Well. It makes sense... but not always.

Here is an counterexample: ` ` (a single space). Actually, a space is a valid file name:
```
$ touch ' '
$ ls -l
total 0
-rw-r--r-- 1 shizh shizh 0 Sep 30 15:10 ' '
```

File: internal/cmd/signer.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+		}
+	}
+
+	// construct signer
+	signer, err := signature.NewSignerFromFiles(keyPath, certPath)
+	if err != nil {
+		return nil, err
+	}
+	if endpoint := ctx.String(FlagTimestamp.Name); endpoint != "" {
+		signer.TSA = timestamp.NewHTTPTimestamper(nil, endpoint)
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply internal/cmd/signer.go:34]: Good point. I will add the check in the `notation-go-lib`.

File: internal/cmd/signer.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/urfave/cli/v2"
+)
+
+// GetSigner returns a signer according to the CLI context.
+func GetSigner(ctx *cli.Context) (notation.Signer, error) {
+	// read paths of the signing key and its corresponding cert.
+	var keyPath, certPath string
+	if path := ctx.String(FlagKeyFile.Name); path != "" {
+		keyPath = path
+		certPath = ctx.String(FlagCertFile.Name)
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply internal/cmd/signer.go:19]: The emptiness will be checked in line 29 by `signature.NewSignerFromFiles`.

File: pkg/config/map.go (hunk 1 above)
Comment (shizhMSFT) [reply pkg/config/map.go:61]: The JSON property names are different for `CertificateMap` and `KeyMap`. Although we can use `interface{}` and have a generic struct, `Marshal` works but `Unmarshal` simply does not work.

File: pkg/signature/jws.go (hunk 2 above)
Comment (shizhMSFT) [reply pkg/signature/jws.go:20]: As mentioned in the previous comment reply, whitespaces are actually valid filenames.

File: pkg/signature/jws.go (hunk 3 above)
Comment (shizhMSFT) [reply pkg/signature/jws.go:64]: `Roots` contains trusted CA certs. We need trusted CA certs to verify the root trust.
`Intermediates` contains untrusted intermediate certs.

File: internal/cmd/signer.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply internal/cmd/signer.go]: @priteshbandi Code updated
- Commit https://github.com/notaryproject/notation-go-lib/pull/11/commits/989490197957d551768c7d73916826e319f8a23b
- Commit fbd5762

**Review by priteshbandi — APPROVED** [review]
LGTM

**Review by SteveLasker — APPROVED** [review]

**Review by sajayantony — APPROVED** [review]

**Pull request merged**